### ISSUE TITLE
Semantic Loss function

### ISSUE THREAD

**Issue by SharhadBashar** [zengxianyu/jsws#9]
In the paper, Lc is defined as follows:
<img width="369" alt="Screen Shot 2020-07-15 at 4 30 30 AM" src="https://user-images.githubusercontent.com/9545714/87522734-f7c9bc80-c653-11ea-8a7f-21bb0c5f6294.png">
In the code, It is:
`loss_cls = F.binary_cross_entropy_with_logits(pred_cls[:, 1:], gt_cls[:, 1:]) + F.binary_cross_entropy_with_logits(pred_cls32x[:, 1:], gt_cls[:, 1:])`
Which one of these are t_i and which one is t_i hat?
Also why are they two loss functions in one? as in what is the difference between pred_cls and pred_cls32x?

**Comment by zengxianyu** [comment]
gt_cls[:, 1:] is t_i. pred_cls32x[:, 1:] and pred_cls[:, 1:] are t_i hat. I apply the loss L_c on multi-scale output so there are two loss. pred_cls is the class scores derived by averaging the 256x256 segmentation output and pre_cls32x is from 8x8 output

**Comment by SharhadBashar** [comment]
follow up: so pred_cls32x[:, 1:] and pred_cls[:, 1:] are t_i hat
specifically, which index in pred_cls is the n in t hat, and which is the t_i?

**Comment by zengxianyu** [comment]
the first dimension is n and the second dimension is i (pred_cls[n, i])

**Comment by SharhadBashar** [comment]
I just wanted to know something, for clarification purposes
if i wanted to replace lines lines 178 and 179: 
`loss_cls = F.binary_cross_entropy_with_logits(pred_cls[:, 1:], gt_cls[:, 1:]) + F.binary_cross_entropy_with_logits(pred_cls32x[:, 1:], gt_cls[:, 1:])`
with 2 for loops, what would that look like?

**Comment by SharhadBashar** [comment]
Also I printed out the shape of pred_cls. why is it (8, 21)? that means its only using 8 images?
Why is the batch size only 8? and will it matter if i change that to include all 12000+ images?

**Comment by zengxianyu** [comment]
> I just wanted to know something, for clarification purposes
> if i wanted to replace lines lines 178 and 179:
> `loss_cls = F.binary_cross_entropy_with_logits(pred_cls[:, 1:], gt_cls[:, 1:]) + F.binary_cross_entropy_with_logits(pred_cls32x[:, 1:], gt_cls[:, 1:])`
> with 2 for loops, what would that look like?

what do you want to iterate over?

**Comment by zengxianyu** [comment]
> Also I printed out the shape of pred_cls. why is it (8, 21)? that means its only using 8 images?
> Why is the batch size only 8? and will it matter if i change that to include all 12000+ images?

there are 8 images in each mini-batch and all the mini-batches cover the whole dataset (e.g., 12000+ images)

**Comment by SharhadBashar** [comment]
pred_cls[:, 1:] is t_i hat
in the paper, t_i hat is a value between [0, 1]
however, when i print out pred_cls[:, 1:], some of the values are negative
how does that happen? and how does the code account for these values? because taking the log of a negative number gives an imaginary result

**Comment by zengxianyu** [comment]
> pred_cls[:, 1:] is t_i hat
> in the paper, t_i hat is a value between [0, 1]
> however, when i print out pred_cls[:, 1:], some of the values are negative
> how does that happen? and how does the code account for these values? because taking the log of a negative number gives an imaginary result

Yes pred_cls[:, 1:] have negative values. sigmoid(pred_cls[:, 1:]) is t_i hat in the paper.
